chia/server/ws_connection.py

```diff
@@ -335,33 +335,37 @@ async def send_messages(self, messages: List[Message]):
         for message in messages:
             await self.outgoing_queue.put(message)
 
+    async def _wait_and_retry(self, msg: Message, queue: asyncio.Queue):
+        try:
+            await asyncio.sleep(1)
+            await queue.put(msg)
+        except Exception as e:
+            self.log.debug(f"Exception {e} while waiting to retry sending rate limited message")
+            return
+
     async def _send_message(self, message: Message):
         encoded: bytes = bytes(message)
         size = len(encoded)
         assert len(encoded) < (2 ** (LENGTH_BYTES * 8))
+        rate_limit_exception_reason = None
+        if is_localhost(self.peer_host):
+            rate_limit_exception_reason = "peer is localhost"
+        # TODO: fix this special case. This function has rate limits which are too low.
+        if ProtocolMessageTypes(message.type) == ProtocolMessageTypes.respond_peers:
+            rate_limit_exception_reason = "we are gossiping peers"
+
         if not self.outbound_rate_limiter.process_msg_and_check(message):
```

Comment:
I think doing it this way is risky. by calling the `process_msg_and_check()` for messages that aren't rate limited, you still increment the message and byte counters that affect messages that *are* rate limited. i.e. it becomes possible for an non-rate limited local peer to starve out all other peers.

I think all the messages and peers that are not subject to rate limiting should circumvent this call entirely.

In fact, if you were to put this logic in the rate limiter itself, it would be easier to unit test (but IMO, the rate limiter is already doing too many things, and should probably be split up into smaller simpler rate limiters)

Reply:
> I think all the messages and peers that are not subject to rate limiting should circumvent this call entirely.

Agreed. This is just a fixup hack to an earlier hack - I don't actually know what is going on in here yet.

Do you want to wait on pulling this in until we get clean byte counts / rate limit separation?

Reply:
I don't think there are any messages that are not rate limited.
And this rate limit object is for this peer only (it's in the ws_connection), so it shouldn't affect other peers.